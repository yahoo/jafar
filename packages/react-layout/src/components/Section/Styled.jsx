/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import styled from 'styled-components';

const sectionSizeCache = {};

export default (size, root) => {
  size = Math.max(size, 0);
  const key = `${size},${root}`;

  if (!sectionSizeCache[key]) {
    sectionSizeCache[key] = {
      Wrapper: styled.div`
        padding: ${() => (root ? `${(size * 8) + 18}px ${(size * 10) + 10}px ${size * 10}px ${(size * 10) + 10}px` : '0')};
        border-bottom: ${() => (root ? '1px solid #d8d8d8' : 'none')};
        &:last-child {
          border-bottom: none;
        }
      `,
      Title: styled.div`
        text-transform: uppercase;
        line-height: 1.1;
        color: inherit;
        font-size: ${() => (14 + (2 * size))}px;
        font-weight: ${() => root ? '500' : '400' };
        padding-bottom: ${() => (Math.max((size * 10) + 10, 24))}px;
        margin: 0;
      `,
    };
  }
  return sectionSizeCache[key];
};

