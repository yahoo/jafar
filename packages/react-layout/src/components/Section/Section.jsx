/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import Box from '../Box';
import Grid from '../Grid';
import Styled from './Styled';

const SectionShape = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
  innerRef: PropTypes.object,
  grid: PropTypes.object,
  boxes: PropTypes.arrayOf(PropTypes.object),
  size: PropTypes.number,
  root: PropTypes.bool,
};
SectionShape.sections = PropTypes.arrayOf(PropTypes.shape(SectionShape));

export class Section extends React.Component {
  static propTypes = SectionShape;

  static defaultProps = {
    size: 4,
    root: true,
    boxes: [],
  };

  render() {
    const { Wrapper, Title } = Styled(this.props.size, this.props.root);

    return (<Wrapper ref={this.props.innerRef} id={this.props.id}>
      {
        this.props.title && <Title>{this.props.title}</Title>
      }
      {
        this.props.grid && <Grid {...this.props.grid} />
      }
      {
        this.props.boxes.map((box, index) => (<Box key={index} { ...box } />))
      }
      {
        (this.props.sections || []).length > 0 && this.props.sections.map(section => (<Section
          key={section.id} size={this.props.size - 1} root={false} {...section} />))
      }
    </Wrapper>);
  }
}

export default React.forwardRef((props, ref) => <Section innerRef={ref} {...props} />);
