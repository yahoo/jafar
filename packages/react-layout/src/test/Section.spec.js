/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import { Section } from '../components/Section/Section';
import SectionWithRef from '../components/Section/index';


describe('Section', () => {
  const Field = () => {};

  it('Should render simple section with ref', () => {
    const section = {
      id: 'job-information',
      boxes: [{
        direction: 'row',
        boxes: getComponentsBoxes(['firstName', 'lastName']),
      }],
      level: 1,
    };
    const component = shallow(<SectionWithRef {...section} />);
    expect(component).toMatchSnapshot();
  });

  it('Should render simple section boxes without title', () => {
    const section = {
      id: 'job-information',
      boxes: [{
        direction: 'row',
        boxes: getComponentsBoxes(['firstName', 'lastName']),
      }],
      level: 1,
    };
    const component = shallow(getComponent(section));
    expect(component).toMatchSnapshot();
  });

  it('Should render simple section boxes - level one', () => {
    const section = {
      id: 'job-information',
      title: 'Job Information',
      boxes: [{
        direction: 'row',
        boxes: getComponentsBoxes(['firstName', 'lastName']),
      }],
      level: 1,
    };
    const component = shallow(getComponent(section));
    expect(component).toMatchSnapshot();
  });

  it('Should render simple section boxes - level two', () => {
    const section = {
      id: 'job-information',
      title: 'Job Information',
      boxes: [{
        direction: 'row',
        boxes: getComponentsBoxes(['firstName', 'lastName']),
      }],
      level: 2,
    };
    const component = shallow(getComponent(section));
    expect(component).toMatchSnapshot();
  });

  it('Should render complex section boxes', () => {
    const section = {
      id: 'job-information',
      title: 'Job Information',
      boxes: [{
        direction: 'row',
        boxes: [{
          direction: 'column',
          boxes: getComponentsBoxes(['firstName', 'lastName']),
        }, {
          direction: 'column',
          boxes: getComponentsBoxes(['address']),
        }],
      }],
    };
    const component = shallow(getComponent(section));
    expect(component).toMatchSnapshot();
  });

  it('Should render section with sub sections', () => {
    const section = {
      id: 'raw-data',
      title: 'Raw Data',
      boxes: [{
        direction: 'row',
        boxes: [{
          direction: 'column',
          boxes: getComponentsBoxes(['id', 'modifier']),
        }, {
          direction: 'column',
          boxes: getComponentsBoxes(['creationDate', 'modificationDate']),
        }],
      }],
      sections: [{
        id: 'raw-data-general',
        title: 'General',
        boxes: [{
          direction: 'row',
          boxes: [{
            direction: 'column',
            boxes: getComponentsBoxes(['id']),
          }, {
            direction: 'column',
            boxes: getComponentsBoxes(['modifier']),
          }],
        }],
      }, {
        id: 'raw-data-modification',
        title: 'Modification',
        boxes: [{
          direction: 'row',
          boxes: [{
            direction: 'column',
            boxes: getComponentsBoxes(['creationDate']),
          }, {
            direction: 'column',
            boxes: getComponentsBoxes(['modificationDate']),
          }],
        }],
      }],
    };
    const component = shallow(getComponent(section));
    expect(component).toMatchSnapshot();
  });

  it('Should render section with only sub sections with level one', () => {
    const section = {
      id: 'raw-data',
      title: 'Raw Data',
      sections: [{
        id: 'raw-data-general',
        title: 'General',
        boxes: [{
          direction: 'row',
          boxes: [{
            direction: 'column',
            boxes: getComponentsBoxes(['id']),
          }, {
            direction: 'column',
            boxes: getComponentsBoxes(['modifier']),
          }],
        }],
        level: 1,
      }, {
        id: 'raw-data-modification',
        title: 'Modification',
        boxes: [{
          direction: 'row',
          boxes: [{
            direction: 'column',
            boxes: getComponentsBoxes(['creationDate']),
          }, {
            direction: 'column',
            boxes: getComponentsBoxes(['modificationDate']),
          }],
        }],
        level: 1,
      }],
    };
    const component = shallow(getComponent(section));
    expect(component).toMatchSnapshot();
  });

  function getComponent(section) {
    return (<Section {...section} />);
  }

  function getComponentsBoxes(fields) {
    return fields.map(id => ({ component: Field, props: { id } }));
  }
});
