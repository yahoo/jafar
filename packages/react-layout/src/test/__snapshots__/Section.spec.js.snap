// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section Should render complex section boxes 1`] = `
<styled.div
  id="job-information"
>
  <styled.div>
    Job Information
  </styled.div>
  <Box
    boxes={
      Array [
        Object {
          "boxes": Array [
            Object {
              "component": [Function],
              "props": Object {
                "id": "firstName",
              },
            },
            Object {
              "component": [Function],
              "props": Object {
                "id": "lastName",
              },
            },
          ],
          "direction": "column",
        },
        Object {
          "boxes": Array [
            Object {
              "component": [Function],
              "props": Object {
                "id": "address",
              },
            },
          ],
          "direction": "column",
        },
      ]
    }
    direction="row"
    key="0"
    props={Object {}}
    style={Object {}}
  />
</styled.div>
`;

exports[`Section Should render section with grid 1`] = `
<styled.div
  id="job-information"
>
  <styled.div>
    Job Information
  </styled.div>
  <CssGrid
    elements={
      Array [
        Object {
          "component": [Function],
          "gridArea": "firstName",
          "props": Object {
            "id": "firstName",
          },
          "selector": "#firstName",
          "style": "width: 350px;",
        },
        Object {
          "component": [Function],
          "gridArea": "lastName",
          "props": Object {
            "id": "lastName",
          },
          "selector": "#lastName",
          "style": "width: 350px;",
        },
        Object {
          "component": [Function],
          "gridArea": "personalId",
          "props": Object {
            "id": "personalId",
          },
          "selector": "#personalId",
          "style": "width: 350px;",
        },
        Object {
          "component": [Function],
          "gridArea": "address",
          "props": Object {
            "id": "address",
          },
          "selector": "#address",
          "style": "width: 350px;",
        },
      ]
    }
    templateAreas={
      Array [
        "firstName lastName .",
        "personalId address .",
      ]
    }
  />
</styled.div>
`;

exports[`Section Should render section with only sub sections with root 1`] = `
<styled.div
  id="raw-data"
>
  <styled.div>
    Raw Data
  </styled.div>
  <Section
    boxes={
      Array [
        Object {
          "boxes": Array [
            Object {
              "boxes": Array [
                Object {
                  "component": [Function],
                  "props": Object {
                    "id": "id",
                  },
                },
              ],
              "direction": "column",
            },
            Object {
              "boxes": Array [
                Object {
                  "component": [Function],
                  "props": Object {
                    "id": "modifier",
                  },
                },
              ],
              "direction": "column",
            },
          ],
          "direction": "row",
        },
      ]
    }
    id="raw-data-general"
    key="raw-data-general"
    root={false}
    size={3}
    title="General"
  />
  <Section
    boxes={
      Array [
        Object {
          "boxes": Array [
            Object {
              "boxes": Array [
                Object {
                  "component": [Function],
                  "props": Object {
                    "id": "creationDate",
                  },
                },
              ],
              "direction": "column",
            },
            Object {
              "boxes": Array [
                Object {
                  "component": [Function],
                  "props": Object {
                    "id": "modificationDate",
                  },
                },
              ],
              "direction": "column",
            },
          ],
          "direction": "row",
        },
      ]
    }
    id="raw-data-modification"
    key="raw-data-modification"
    root={false}
    size={3}
    title="Modification"
  />
</styled.div>
`;

exports[`Section Should render section with sub sections 1`] = `
<styled.div
  id="raw-data"
>
  <styled.div>
    Raw Data
  </styled.div>
  <Box
    boxes={
      Array [
        Object {
          "boxes": Array [
            Object {
              "component": [Function],
              "props": Object {
                "id": "id",
              },
            },
            Object {
              "component": [Function],
              "props": Object {
                "id": "modifier",
              },
            },
          ],
          "direction": "column",
        },
        Object {
          "boxes": Array [
            Object {
              "component": [Function],
              "props": Object {
                "id": "creationDate",
              },
            },
            Object {
              "component": [Function],
              "props": Object {
                "id": "modificationDate",
              },
            },
          ],
          "direction": "column",
        },
      ]
    }
    direction="row"
    key="0"
    props={Object {}}
    style={Object {}}
  />
  <Section
    boxes={
      Array [
        Object {
          "boxes": Array [
            Object {
              "boxes": Array [
                Object {
                  "component": [Function],
                  "props": Object {
                    "id": "id",
                  },
                },
              ],
              "direction": "column",
            },
            Object {
              "boxes": Array [
                Object {
                  "component": [Function],
                  "props": Object {
                    "id": "modifier",
                  },
                },
              ],
              "direction": "column",
            },
          ],
          "direction": "row",
        },
      ]
    }
    id="raw-data-general"
    key="raw-data-general"
    root={false}
    size={3}
    title="General"
  />
  <Section
    boxes={
      Array [
        Object {
          "boxes": Array [
            Object {
              "boxes": Array [
                Object {
                  "component": [Function],
                  "props": Object {
                    "id": "creationDate",
                  },
                },
              ],
              "direction": "column",
            },
            Object {
              "boxes": Array [
                Object {
                  "component": [Function],
                  "props": Object {
                    "id": "modificationDate",
                  },
                },
              ],
              "direction": "column",
            },
          ],
          "direction": "row",
        },
      ]
    }
    id="raw-data-modification"
    key="raw-data-modification"
    root={false}
    size={3}
    title="Modification"
  />
</styled.div>
`;

exports[`Section Should render simple section boxes - not root 1`] = `
<styled.div
  id="job-information"
>
  <styled.div>
    Job Information
  </styled.div>
  <Box
    boxes={
      Array [
        Object {
          "component": [Function],
          "props": Object {
            "id": "firstName",
          },
        },
        Object {
          "component": [Function],
          "props": Object {
            "id": "lastName",
          },
        },
      ]
    }
    direction="row"
    key="0"
    props={Object {}}
    style={Object {}}
  />
</styled.div>
`;

exports[`Section Should render simple section boxes - root 1`] = `
<styled.div
  id="job-information"
>
  <styled.div>
    Job Information
  </styled.div>
  <Box
    boxes={
      Array [
        Object {
          "component": [Function],
          "props": Object {
            "id": "firstName",
          },
        },
        Object {
          "component": [Function],
          "props": Object {
            "id": "lastName",
          },
        },
      ]
    }
    direction="row"
    key="0"
    props={Object {}}
    style={Object {}}
  />
</styled.div>
`;

exports[`Section Should render simple section boxes without title 1`] = `
<styled.div
  id="job-information"
>
  <Box
    boxes={
      Array [
        Object {
          "component": [Function],
          "props": Object {
            "id": "firstName",
          },
        },
        Object {
          "component": [Function],
          "props": Object {
            "id": "lastName",
          },
        },
      ]
    }
    direction="row"
    key="0"
    props={Object {}}
    style={Object {}}
  />
</styled.div>
`;

exports[`Section Should render simple section with ref 1`] = `
<Section
  boxes={
    Array [
      Object {
        "boxes": Array [
          Object {
            "component": [Function],
            "props": Object {
              "id": "firstName",
            },
          },
          Object {
            "component": [Function],
            "props": Object {
              "id": "lastName",
            },
          },
        ],
        "direction": "row",
      },
    ]
  }
  id="job-information"
  innerRef={null}
  root={true}
  size={4}
/>
`;
